### DISCUSSION TITLE
Vendure Security

### DISCUSSION THREAD

**Discussion by cloudcompute** [vendurehq/vendure#792] [General]
Great.. if somebody can answer these questions:

Is Vendure taking care of csrf, xss, and other security attacks?

Vendure is not using JWT strategy for Authentication. Then how is it accomplished. A few lines of explanation will help a lot.

Request for the rate-limit feature.

Where are the sessions stored by default? Is it possible to store the sessions in redis?

**Comment by michaelbromley** [comment]
> Is Vendure taking care of csrf, xss, and other security attacks?

There's no specific mitigations against CSFR or XSS. I plan to revisit this when I work more on https://github.com/vendure-ecommerce/vendure/issues/134 before the first stable release. 

Regarding "other security attacks" I find this a bit too broad to easily answer.

> Vendure is not using JWT strategy for Authentication. Then how is it accomplished. A few lines of explanation will help a lot.

Vendure uses session tokens (a token is just a random string of characters) which are passed in to API calls as a cookie or a header. This token is checked against the list of active Sessions (by default stored in the main DB) to establish whether the call is part of an authenticated session. More here: https://www.vendure.io/docs/storefront/managing-sessions/

Authentication itself is (by default) done using username/password. The password is stored in the DB, hashed using bcrypt. You can also define other auth strategies: https://www.vendure.io/docs/developer-guide/authentication/

> Request for the rate-limit feature.

You can pass any kind of Express middleware in your Vendure config to implement rate-limiting (e.g. https://www.npmjs.com/package/express-rate-limit). You can also do it at other levels of the server stack, e.g. in nginx.

> Where are the sessions stored by default? Is it possible to store the sessions in redis?

In the database. You can define your own strategy if you want to use redis: https://www.vendure.io/docs/typescript-api/auth/session-cache-strategy/

**Comment by cloudcompute** [comment]
Thanks for the reply. The project looks awesome. Kindly answer few more...

a. What is the expected planned release date?

b. As you said above.. this means that you are using Session tokens for dual purposes a. authentication strategy and b. session tracking. This theory was typically used before the advent of JWT, am I right?

c. I hope that you are making sure to remove a session token from the db storage as soon as a user logouts. This makes sure that even if a session cookie is stolen by someone, he cannot use it anymore after the original user logs out.

d. A single user might be having multiple sessions at a same time, for example, a. accessing Vendure API from two web browsers .. or b. mobile device and web browser. Is Vendure taking care of it i.e. ability to maintain 1 to many relationship between User and Session?

For csrf, xss.. these packages could be of help, I am not sure though.
https://www.npmjs.com/package/csurf
https://www.npmjs.com/package/xss

**Comment by michaelbromley** [comment]
> a. What is the expected planned release date?

In the next month or so.

> b. As you said above.. this means that you are using Session tokens for dual purposes a. authentication strategy and b. session tracking. This theory was typically used before the advent of JWT, am I right?

It is still widely used. JWT attempts to solve a related but different problem. In fact, I did start out using JWT but moved away from it for the reasons detailed in this issue: https://github.com/vendure-ecommerce/vendure/issues/25

> c. I hope that you are making sure to remove a session token from the db storage as soon as a user logouts. This makes sure that even if a session cookie is stolen by someone, he cannot use it anymore after the original user logs out.

Indeed. See https://github.com/vendure-ecommerce/vendure/blob/532ea21432b29bda59baa6116531cbcc00493298/packages/core/src/api/resolvers/base/base-auth.resolver.ts#L73

> Is Vendure taking care of it i.e. ability to maintain 1 to many rrelationship between User and Session?

Yes: https://github.com/vendure-ecommerce/vendure/blob/e737c56ad7db2797fafb1e19e0d662da5de0be82/packages/core/src/entity/session/authenticated-session.entity.ts#L20-L26

Thanks for the links to the packages. I'll research those in due time 👍

**Comment by cloudcompute** [comment]
Very satisfactory answers. Looking at the source code, very well structured.

In this link, what do you want to convey.. when you said >No this won't work - it removes the ability to invalidate sessions. Once a cookie of this type has been issued with an expiry of, e.g. 2 weeks, then there would be no way to revoke access for that user.
https://github.com/vendure-ecommerce/vendure/issues/25#issuecomment-423980232

Access to session data is frequently used in an ecommerce site, like for example, wishlist, cart, product comparison. And use of redis being the fastest is recommended over the db. Also redis easily scalable. In this scenario, redis should probably be the default for session.

Another point related to security.. Is Vendure making sure that internal errors and the corresponding error codes of the db and Apollo Graphql  are not sent in the response body to the frontend in the production environment? Instead they should be logged in a logger file. Yes, during development, a developer would like to see the entire stack information including the db/graphql-level errors and their codes in the frontend screen itself. But during production, may be an error code defined at the Vendure application level is sufficient, for example, something like this ... "Looks like there is an internal server error whose error code is 021. Kindly inform the system developers. about this error."

To implement this, we may need to map the application-level error codes with the db and graphql errors in the 'error-constants' file. What is your viewpoint?

**Reply by michaelbromley** [reply]
> it removes the ability to invalidate sessions

I think I meant that encoding any data into a session token leads to the problem of stale data. And in the end there is no benefit to doing this if we are already performing a DB query for the Session on each request anyway.

> redis should probably be the default for session.

I don't want to make redis a requirement for running Vendure. My approach is that the core should be as simple and dependency-free as possible, and then allow extension as needed. This is also why I don't use Elasticsearch as the default search method, even though most larger projects will probably want to use it. And why I use the DB as the basis of the default job queue rather than Redis, RabbitMQ etc. For many small-to-medium projects, the DB is fine for all of these.

What _would_ make sense is to supple a redis-based SessionCacheStrategy as a plugin. I've just not had time to do that so far.

Re. exposing internal errors, there might be some improvements to make there. Also bear in mind that the storefront app that the final customer deals with can choose to display errors as it likes.

**Comment by cloudcompute** [comment]
Wonderful explanation. Thanks for your time.

**Reply by kevmtt** [reply]
Hey @ramandhingra
All I know about redis is that its an in-memory key-value database. Can you explain me why you think it should be default for session management? 

_I'm trying to build a mobile application (where a user should be signed in as long as possible)._

**Reply by cloudcompute** [reply]
Hello @Draykee

The main reason is all the session-related read-write operations are carried out in memory and therefore very fast. Pl. go through this 'Stack Overflow' answer
https://stackoverflow.com/a/10279573

You may also read this..
https://redislabs.com/blog/cache-vs-session-store/
